### PR TITLE
docs: Add architecture overview

### LINKED ISSUE #2791: Add architecture overview document
We should add an architecture overview document (under `docs`) documenting all the components used in the Oasis Core implementation and a high-level description of their interactions.

### LINKED ISSUE #2836: Add client API documentation
We should add client API documentation for the gRPC services exposed via the internal oasis-node socket.

### PR DESCRIPTION
**Pull request by kostko**
Fixes #2791 
Fixes #2836 

[Rendered version](https://github.com/oasislabs/oasis-core/blob/kostko/feature/docs-arch-overview/docs/index.md)

### CODE REVIEW

File: docs/oasis-node/rpc.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+[consensus]: ../consensus/index.md
+[runtime]: ../runtime/index.md
+[submit transactions]: ../consensus/transactions.md#submission
+
+## Protocol
+
+As other parts of Oasis Core, the RPC interface exposed by Oasis Node uses the
+[gRPC protocol] with the CBOR codec (instead of Protocol Buffers). If your
+application is written in Go, you can use the convenience gRPC wrappers provided
```
Comment (ptrus) [inline docs/oasis-node/rpc.md:19]: Could also link to our rust client crate as an example of a rust gRPC wrapper? (although with a note that it only implements some of the methods). Maybe [client/src/transaction/api/client.rs](https://github.com/oasislabs/oasis-core/blob/master/client/src/transaction/api/client.rs) is a good example

File: docs/runtime/index.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+verify that all nodes have indeed computed the same result. As mentioned in case
+of discrepancies it will instruct nodes elected as _backups_ to repeat the
+computation.
+
+[cryptographic commitments]: ../../go/roothash/api/commitment
+
+### Storage Receipts
+
+All runtime persistent state is stored by storage nodes. These provide a
+[Merklized Key-Value Store (MKVS)] to compute nodes. The MKVS provides immutable
```
Comment (ptrus) [inline docs/runtime/index.md:150]: This is not rendered as a link (probably cause (MKVS) is missing bellow)

File: docs/oasis-node/rpc.md (hunk 1 above)
Comment (kostko) [reply docs/oasis-node/rpc.md:19]: Good idea, added.

File: docs/consensus/index.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+- [Epoch Time], an epoch-based time keeping service.
+- [Random Beacon], a source of randomness for other services.
+- [Staking], operations required to operate a PoS blockchain.
+- [Registry], an entity/node/runtime public key and metadata registry service.
+- [Committee Scheduler] service.
+- [Root Hash], runtime commitment processing and minimal runtime state keeping
+  service.
+- [Key Manager] policy state keeping service.
+
+Each of the above services provides methods to query the current state. In order
```
Comment (peterjgilbert) [inline docs/consensus/index.md:28]:
```suggestion
Each of the above services provides methods to query its current state. In order
```

File: docs/consensus/index.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+to mutate the current state, each operation needs to be wrapped into a
+[consensus transaction] and submitted to the consensus layer for processing.
+
+Oasis Core defines an interface for each kind of service (in `go/<service>/api`)
```
Comment (peterjgilbert) [inline docs/consensus/index.md:32]:
```suggestion
Oasis Core defines an interface for each kind of service (in `go/<service>/api`),
```

File: docs/consensus/index.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+<!-- markdownlint-enable line-length -->
+
+## Tendermint
+
+![Tendermint](../images/oasis-core-consensus-tendermint.svg)
+
+The Tendermint consensus backend lives in [`go/consensus/tendermint`].
+
+For more information about Tendermint itself see
+[the Tendermint Core developer documentation], this section assumes familiarity
```
Comment (peterjgilbert) [inline docs/consensus/index.md:60]:
```suggestion
[the Tendermint Core developer documentation]. This section assumes familiarity
```

File: docs/consensus/index.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+1. The second part is the **query and transaction submission glue** that makes
+   it easy to interact with the ABCI application, presenting everything via the
+   Oasis Core Consensus interface.
+
+[`go/consensus/tendermint`]: ../../go/consensus/tendermint
+[the Tendermint Core developer documentation]: https://docs.tendermint.com/
+
+### ABCI Application Multiplexer
+
+Tendermint Core consumes consensus layer logic via the [ABCI protocol] which
```
Comment (peterjgilbert) [inline docs/consensus/index.md:80]:
```suggestion
Tendermint Core consumes consensus layer logic via the [ABCI protocol], which
```

File: docs/consensus/keymanager.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
+# Key Manager
+
+The key manager service is responsible for coordinating the SGX-based key
+manager runtimes. It stores and publishes policy documents and status updates
```
Comment (peterjgilbert) [inline docs/consensus/keymanager.md:4]: At a minimum we should say what policies cover: which enclaves are allowed to fetch which keys, and which enclaves the master secret can be replicated to?

File: docs/consensus/registry.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
+# Registry
+
+The registry service is responsible for managing an entity/node/runtime public
+key and metadata registry.
```
Comment (peterjgilbert) [inline docs/consensus/registry.md:4]:
```suggestion
The registry service is responsible for managing a registry of entity/node/runtime
public keys and metadata.
```

File: docs/consensus/registry.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+<!-- markdownlint-disable line-length -->
+[`go/registry/api`]: ../../go/registry/api
+[consensus service API documentation]: https://pkg.go.dev/github.com/oasislabs/oasis-core/go/registry/api?tab=doc
+<!-- markdownlint-enable line-length -->
+
+## Methods
+
+### Register Entity
+
+Entity registration enables a new entity to be created. A new register entity
```
Comment (peterjgilbert) [inline docs/consensus/registry.md:19]: We should define somewhere (probably not here) what an "entity" is. This text should probably link to that definition.

Comment (peterjgilbert) [reply docs/consensus/registry.md:19]: For reference, this is what we have in the operator docs: https://docs.oasis.dev/operators/architecture-overview.html#entities-and-key-management. It's a bit high-level/casual, matching the overall tone of those docs.

File: docs/consensus/registry.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+}
+```
+
+**Fields:**
+
+* `node_id` specifies the node identifier of the node to thaw.
+
+The transaction signer MUST be the entity key that owns the node.
+
+Thawing a node requires that the node's freeze period has already passed.
```
Comment (peterjgilbert) [inline docs/consensus/registry.md:125]: Might be useful to note that the duration of the freeze period for a given event (e.g., double signing) is a consensus parameter.

File: docs/consensus/registry.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+```
+registry.RegisterRuntime
+```
+
+The body of a register runtime transaction must be a [`SignedRuntime`]
+structure, which is a [signed envelope] containing a [`Runtime`] descriptor.
+The signer of the transaction MUST be the owning entity key.
+
+Registering a runtime may require sufficient stake in the owning entity's
+[escrow account].
```
Comment (peterjgilbert) [inline docs/consensus/registry.md:147]: Might be useful to point out the consensus parameter that defines this threshold.

File: docs/consensus/transactions.md [hunk 12]
```diff
@@ -1,6 +1,17 @@
 # Transactions
 
-The consensus layer uses a common transaction format for all transactions.
+The consensus layer uses a common transaction format for all transactions. As
+everything else it tries to be independent of any concrete [consensus backend].
```
Comment (peterjgilbert) [inline docs/consensus/transactions.md:4]:
```suggestion
with other Oasis Core components, it tries to be independent of any concrete [consensus backend].
```

File: docs/consensus/transactions.md [hunk 13]
```diff
@@ -80,3 +90,34 @@ Fields:
 
 * `amount` is the total fee amount to be paid.
 * `gas` is the maximum gas that an operation can use.
+
+## Gas Estimation
+
+As transactions need to include the maximum amount of gas that can be consumed
```
Comment (peterjgilbert) [inline docs/consensus/transactions.md:96]:
```suggestion
As transactions need to provide the maximum amount of gas that can be consumed
```

**Comment by peterjgilbert** [comment]
I added a bunch of minor suggestions and a few comments about things that should probably be further explained. Looks great overall!

File: docs/runtime/index.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+In case of a discrepancy, the computation must be repeated using a separate
+larger compute committee which decides what the correct results were. Since all
+commitments are attributable to compute nodes, any node(s) that produced
+incorrect results may be subject to having their stake slashed and may be
+removed from future committees.
+
+Given the above, an additional constraint with replicated runtimes is that they
+must be fully deterministic, meaning that a computation operating on the same
+initial state executing the same inputs (transactions) must always produce the
+same outputs and new state.
```
[the remark this answers is not in the document]
Comment (kostko) [reply docs/runtime/index.md:119]: Good idea.

File: docs/consensus/registry.md (hunk 9 above)
Comment (kostko) [reply docs/consensus/registry.md:19]: Added.

File: docs/consensus/keymanager.md (hunk 7 above)
Comment (kostko) [reply docs/consensus/keymanager.md:4]: Added.

**Pull request merged**